File: Uebung 9/Ising.f90

Program Ising

  external CalcE

  integer NMax,NN                  !/* Maximal lattice size, # neighbors*/
  real*8 JJ, HH                      !/* Coupling constant, magnetic field */
  parameter (NMax=50,NN=4)
  parameter (JJ=1.d0,HH=0.d0)


  integer mag(Nmax,Nmax)            !/* 2D Ising Lattice */
  integer i, j, k                   !/* Loop counters */
  integer s,d                       !/* Lattice spin variables */  
  real*8 Energy                     !/* Total lattice energy */
  integer Inn(NN)                    !/*  Nearest neighbor array I  */
  parameter (Inn=(/1,-1,0,0/))
  integer Jnn(NN)                       !/*  Nearest neighbor array J  */
  parameter (Jnn=(/0,0,1,-1/)) 
  integer Inew, Jnew                !/* Nearest neighbot indices */ 
  real*8 Etemp, deltaE              !/* Temp energy variables for MC moves */ 
  integer accept                    !/* Number of accepted moves */   
  integer move                      !/* Number of moves total */   

  real*8 T                          !/* temperature (in units of J/k_B) */
  real*8 beta
  integer sweeps                    !/* number of measurement sweeps */
  integer warm                      !/* number of warm-up sweeps */
  integer L                             !/* lattice dimension */


 ! /***************************
 !  * Initialization          *
 !  ***************************/

  write (*,*) "Temperatur ?"
  read(*,*) T
  beta=1.d0/T

  write (*,*) "Gittergroesse (L x L) ?"
  read(*,*) L

  write (*,*) "# sweeps ?"
  read(*,*) sweeps
  
  write (*,*) "# warm up sweeps?"
  read(*,*) warm

  call random_init(mag,L,Nmax)  
  Energy = CalcE(mag,L,Jnn,Inn,NN,JJ,HH,Nmax)

 
! /***************************
! /* warum up sweeps */
! /***************************
 
 
 
! /***************************
! /* end warum up sweeps */ 
! /***************************
 
  



 

! /***************************
! /*  sweeps */
! /***************************



! /***************************
! /*  end sweeps */ 
! /***************************

  call  outputmag(mag,L,Nmax)
  write (*,*) 'Average energy',Energy
end Program Ising







subroutine random_init(mag,L,Nmax)
  integer Nmax
  integer mag(Nmax,Nmax), L
 
  integer i,j
  
  do i=1,L
     do j=1,L
       	if (rand().gt.0.5d0) then
	   mag(i,j) = 1
        else
           mag(i,j) = -1
        endif
     enddo
  enddo

  return
end subroutine random_init



subroutine outputmag(mag,L,Nmax)
  integer Nmax
  integer mag(Nmax,Nmax), L
 
  integer i,j
  write (*,*) 'Spin snapshot'

  do i=1,L
     do j=1,L
        if(mag(i,j).eq. 1) then
	   write(*,'(a)',ADVANCE='NO') '-> '
        else
           write(*,'(a)',ADVANCE='NO') '<- '
        endif
      
     enddo
     !     Newline to complete
     write (*,*) 
     write (*,*)
  enddo
  return 
end subroutine outputmag




real function CalcE(mag,L,Jnn,Inn,NN,JJ,HH,Nmax)
  integer Nmax
  integer mag(Nmax,Nmax),Jnn(NN),Inn(NN),L,NN
  
  real*8 JJ, HH
  integer i,j,k,Inew,Jnew
  real*8 Energy

  !/* Determine the initial energy */
  Energy = 0.0
  
  do i=1,L
     do j=1,L
        
	!/* Loop over nearest neighbors */
	do k=1, NN  
            Inew = i + Inn(k)       
            Jnew = j + Jnn(k)
    
	    !/* Check periodic boundary conditions */
	    if (Inew .le. 0) then
	      Inew = L
            else 
               if(Inew .gt. L) then 
                  Inew = 1
               endif
            endif    
            if (Jnew .le. 0) then
	      Jnew = L
            else 
               if(Jnew .gt. L) then 
                  Jnew = 1
               endif
            endif
	    
	    !/* Update the energy */
	    Energy = Energy-JJ * mag(i,j) * mag(Inew,Jnew)
        enddo
	!/*Calculate the contribution from the field H */
	Energy = Energy - 2.d0*HH*mag(i,j);
     enddo
   enddo

   !/* Account for double counting */
   Energy = Energy/2.d0;

   CalcE=Energy
   return
 end function CalcE

File: Uebung 9/Ising.c
#define PROGNAME "mc_Ising_2D"
#define VERSION "0.2a"
#define DATE  "today"
#define AUTHOR "you"

/* Metropolis Monte Carlo for 2D-Ising Model with periodic boundary conditions */
/* Code adapted from http://gold.cchem.berkeley.edu/~acpan/220A/2DIsing.c and N. Bluemer*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

/*************************
 * Constant Declarations *
 ************************/
#define NMAX 20            /* Lattice size (i.e. NMAX x NMAX) */

#define J 1.0             /* Ising lattice coupling constant 
			    J > 0: Ferromagnetic 
			    J < 0: Antiferromagnetic */
#define H 0.0              /* Ising lattice field strength */

void error(char error_text[])
/* standard error handler */
{
    fprintf(stderr,"\n%s run-time error\n",PROGNAME);
    fprintf(stderr,"--%s--\n",error_text);
    fprintf(stderr,"for general help use option -h\n");
    fprintf(stderr,"...now exiting to system...\n");
    exit(1);
}

void random_init(int mag[NMAX][NMAX], int L)
{
  int i,j;
  
  for(i = 0; i < L; i++)
    for(j = 0; j < L; j++)
      {
	if(drand48() > 0.5)
	  mag[i][j] = 1;
	else
	  mag[i][j] = -1;
      }
}

void outputmag(int mag[NMAX][NMAX], int L)
{
  int i,j;

  printf("\n");

  for(i = 0; i < L; i++)
    {
      for(j = 0; j < L; j++)
	{
	  if(mag[i][j] == -1)
	    printf("* ");
	  else
	    printf("  ");
	}
      printf("\n");
    }
}

double mag_av(int mag[NMAX][NMAX], int L)
/* computes average magnetization */
{
  int i,j;
  double x;

  x=0.0;

  for(i = 0; i < L; i++)
    for(j = 0; j < L; j++)
      x+=1.0*mag[i][j];
  x=x/(L*L);
  return(x);
}

void printhelp ()
{
 printf("**********************************************************\n");
 printf("%s: Metropolis Monte Carlo for 2D Ising model\n",PROGNAME);
 printf("Version: %s, %s by %s\n",VERSION,DATE,AUTHOR);
 printf("based on http://gold.cchem.berkeley.edu/~acpan/220A/2DIsing.c\n");
 printf("options: -T temperature (really: k_B T/|J|)\n");
 printf("         -L linear dimension of lattice (L<=%d)\n",NMAX);
 printf("         -w# number of initialization sweeps\n");
 printf("         -n# number of measurement sweeps\n");
 printf("         -c print configurations\n");
 printf("         -h this help\n");
}

/***********************************************************/
int main(int argc, char *argv[])
{
  /*************************
   * Variable Declarations *
   ************************/
  int mag[NMAX][NMAX];           /* 2D Ising Lattice */
  int i, j, k;                   /* Loop counters */
  int s,d;                       /* Lattice spin variables */  
  double Energy;                 /* Total lattice energy */
  int Inn[8] = {1,-1,0,0,1,1,-1,-1};       /* Nearest neighbor array I */
  int Jnn[8] = {0,0,1,-1,1,-1,1,-1};       /* Nearest neighbor array J */
  int Inew, Jnew;                /* Nearest neighbot indices */ 
  double Etemp, deltaE;          /* Temp energy variables for MC moves */ 
  int accept = 0;                /* Number of accepted moves */
  int move = 0;                  /* Number of moves total */ 

  char c;
  double T;                      /* temperature (in units of J/k_B) */
  double beta;
  int sweeps;                    /* number of measurement sweeps */
  int warm;                      /* number of warm-up sweeps */
  int print_conf;                /* flag for printing configurations */
  int L;                         /* lattice dimension */


  /***************************
   * Initialization          *
   ***************************/
  T=1.5;
  sweeps=100;
  print_conf=0;
  L=NMAX;

  while (--argc > 0 && (*++argv)[0] == '-')
    while (c= *++argv[0])
	   switch (c) {
	   case 'n':
 	     sscanf(++argv[0],"%d\n",&sweeps); 
             break;
	   case 'w':
 	     sscanf(++argv[0],"%d\n",&warm); 
             break;
	   case 'L':
 	     sscanf(++argv[0],"%d\n",&L); 
             if (L>NMAX) error ("L too large");
             break;
           case 'T':
 	     sscanf(++argv[0],"%lf\n",&T); 
             break;
           case 'c':
             print_conf=1;
             break;
	   case 'h':
	     printhelp();
	     exit(0);
             break;
/*  	   default:  */
/*  	     error("No valid choice");  */
	   }

  beta=1.0/T;
  printf("# L=%d, T=%f, n=%d, print_conf=%d\n", L, T, sweeps, print_conf);
  printf("# magnetization/site, energy/site\n");

  /* Seed the random number generator */
  srand48((unsigned int) time(NULL) - (time(NULL)/100000)*100000);

  /* Generate random initial configuration */
  random_init(mag, L);

  /* Determine the energy initially */
  Energy = 0.0;
  
  for(i=0;i<L;i++)
    for(j=0;j<L;j++)
      {
	/* Loop over nearest neighbors */
	for(k=0;k<8;k++)  
	  {
	    Inew = i + Inn[k];       
	    Jnew = j + Jnn[k];
	  
	    /* Check periodic boundary conditions */
	    if(Inew < 0)
	      Inew = L-1;
	    else if(Inew >= L)
	      Inew = 0;
	    
	    if(Jnew < 0)
	      Jnew = L-1;
	    else if(Jnew >= L)
	      Jnew = 0;
	    
	    /* Update the energy */
	    Energy += -J * mag[i][j] * mag[Inew][Jnew];
	  }
	/*Calculate the contribution from the field H */
	Energy += 2*H*mag[i][j];
      }
  
  /* Account for double counting */
  Energy /= 2.0;
	
  /* Monte Carlo Initialization */
  for(i=0;i<warm;i++)
    {
      for(j=0;j<L*L;j++)
	{
    /* ToDO: put Metropolis update etc here !!!!!!!!!!!!!!!!!!! */
	}
    }
  
  /*************************** 
   * The Run                 * 
   ***************************/
  for(i=0;i<sweeps;i++){

    /* ToDo: put Metropolis update etc here !!!!!!!!!!!!!!!!!!! */

      /* COLLECT DATA HERE */
      printf ("%9.6lf  %9.6lf\n", mag_av(mag,L), Energy/(L*L));
    }

  /*****************************
   * Output data and shut down *                 
   ****************************/

  /* OUTPUT DATA HERE */
  if (print_conf>0)
    outputmag(mag, L);

  /*Ratio of accepted spin flips compared to total attempted flips*/
  printf("# Acceptance rate: %2.1f %%\n", 100.0*accept/move);
}

